core/src/main/java/org/elasticsearch/common/xcontent/ObjectParser.java

```diff
@@ -395,8 +453,12 @@ private void parseSub(XContentParser parser, FieldParser fieldParser, String cur
 
     private FieldParser getParser(String fieldName) {
         FieldParser parser = fieldParserMap.get(fieldName);
-        if (parser == null && false == ignoreUnknownFields) {
-            throw new IllegalArgumentException("[" + name  + "] unknown field [" + fieldName + "], parser not found");
+        if (parser == null) {
+            if (this.unknownFieldParser != null) {
+                parser = this.unknownFieldParser;
+            } else if (false == this.ignoreUnknownFields) {
+                throw new IllegalArgumentException("[" + name  + "] unknown field [" + fieldName + "], parser not found");
+            }
```

Comment:
I find it confusing to have a fallback/unknown field parser and still be able to ignore unknown fields. I'd make these two mutually exclusive, ie not possible to define an unknown field parser if `ignoreUnknownFields` is set to true.